test/tests/parallel/set3/mta/dsn.bats

```diff
@@ -0,0 +1,95 @@
+load "${REPOSITORY_ROOT}/test/helper/setup"
+load "${REPOSITORY_ROOT}/test/helper/common"
+
+BATS_TEST_NAME_PREFIX='[DSN] '
+CONTAINER1_NAME='dms-test_dsn_send_always'
+CONTAINER2_NAME='dms-test_dsn_send_auth'
+CONTAINER3_NAME='dms-test_dsn_send_none'
+# A similar line is added to the log when a DSN (Delivery Status Notification) is sent:
+#
+# postfix/bounce[1023]: C943BA6B46: sender delivery status notification: DBF86A6B4CO
+#
+LOG_DSN='delivery status notification'
+
+function setup_file() {
+  local CUSTOM_SETUP_ARGUMENTS=(
+    # Required only for delivery via nc (_send_email)
+    --env PERMIT_DOCKER=container
+  )
+
+  export CONTAINER_NAME=${CONTAINER1_NAME}
+  _init_with_defaults
+  # Unset `smtpd_discard_ehlo_keywords` to allow DSNs by default on any `smtpd` service:
+  cp "${TEST_TMP_CONFIG}/dsn/postfix-main.cf" "${TEST_TMP_CONFIG}/postfix-main.cf"
+  _common_container_setup 'CUSTOM_SETUP_ARGUMENTS'
+  _wait_for_service postfix
+  _wait_for_smtp_port_in_container
```

Comment:
You could move these wait lines into each test case.

Not required, I'm fine with this as-is; I just thought I'd mention it (_slight benefit described below_).

---

The first test won't run until all 3 containers start sequentially with these blocking calls. Probably only a few seconds, so I'm not too fussed.

Some test files you'll notice handle this in each test case, as you had done before, which uses the simple `teardown()`. Lower memory overhead as well, but I don't think it's that much for a minimal container setup.

Reply:
If it's important I could change it, just not at the moment as it's a bit late and if I change it I'd want to also run some tests to make sure it doesn't break anything.

If you want to merge earlier, you can also change it if find have time, it should not be a problematic change.

Reply:
I don't mind, it's an easy optimization to spot and address later if someone cares to do that.

We'll wait for @georglauterbach to chime in, if they approve then this can be merged as-is :)

Reply:
I tested it just to be sure:

`_wait_for_*` in `setup_file()`
```
real	0m14.120s
user	0m1.006s
sys	0m0.434s
```

`_wait_for_*` in `@test`
```
real	0m14.030s
user	0m0.987s
sys	0m0.493s
```

Reply:
Tested how? The concern was about all 3 being setup in `setup_file()`, this means:
- Container 1 start and wait.
- Container 2 start and wait after Container 1 is ready.
- Container 3 start and wait after Container 2 is ready.

There is a staggered wait sequence as a result. The waits could all be handled at the end of `setup_file()` which would minimize that. You'd still have a bit more delay / resources used spinning up multiple container instances, rather than one at a time per test-case (_since each test case has it's own container, rather than multiple test cases that can share the same container config_).

The wait time may vary a bit (_probably not by much, but can be a couple seconds IIRC, as the inconsistent test failures showed earlier_).

Reply:
No need to change this, it looks good. I may go over it later since the review feedback already was quite extensive. 👍🏼